layerscape: add Traverse LS1043-S board

mcbridematt opened the pull request:
The Traverse LS1043-S board is a router board based on NXP/Freescale's LS1043 SoC, with 4x1GBase-T, 1 SFP and 1 SFP+, as well as 2xminiPCIe (for dual band WiFi), M.2/B LTE and M.2/M SATA (via USB bridge).

Unlike the Layerscape reference boards, the LS1043S board has NAND flash and uses the mainline U-Boot.

This patchset implements support for the LS1043-S board, as well as the earlier LS1043-V board (which was sampled, but not produced). The FIT image generated is a multi-configuration image
that has DTBs embedded for both boards.

The target name is set to 'traverse-five64' as that product will become the 'flagship' for this range.
It is our intention that all boards in this family (LS1043-S and later, Five64) will boot the same binary.

Not included in this patch are the hwmon drivers not in the kernel (emc1704,pac1934), bootloader or "rescue" SD card creation, which will be added at a later date, pending upstream.

We also have a version of the layerscape target with kernel 4.14 operational, but I can't vouch for other Layerscape platforms (LS1012,1046,LS1088 etc.)

Product Page: https://traverse.com.au/products/ls1043s-router-board/
Manual: https://traverse.com.au/docs/ls1043s/
Our repositories: https://gitlab.com/traversetech

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
layerscape: remove DEVICE_TYPE=developerboard

The Traverse LS1043 series are router boards and need the default router package
selection

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
layerscape: enable QorIQ thermal driver

This provides access to the CPU die temperature sensor and
the thermal throttling integration

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
layerscape: add support for fixed-link on RGMII interfaces

This is required on the Traverse LS1043 boards to support SFP
and xDSL plug-ins.

This will not be needed on kernel 4.14.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
uboot-envtools: add configuration for Traverse LS1043 boards.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

pprindeville commented on package/boot/uboot-envtools/files/layerscape:
```diff
@@ -0,0 +1,25 @@
+# Copyright (C) 2016 LEDE
+#
+
+[ -e /etc/config/ubootenv ] && exit 0
```
I would `-f` instead...

pprindeville commented on package/boot/uboot-envtools/files/layerscape:
```diff
@@ -0,0 +1,25 @@
+	traverse,ls1043v | \
+	traverse,ls1043s)
+		ubootenv_add_uci_config "/dev/mtd1" "0x40000" "0x2000"  "0x20000"
+	;;
```
Not sure I understand the lack of a default case...  do you want to proceed even when you don't match either of those two?

mcbridematt replied:
Good question, the structure here is exactly the same as the other archs in uboot-envtools/files/

I think the intention was that a config for an unknown board in UCI will be retained (by "config_foreach ubootenv_add_app_config ubootenv" underneath), that said, I'm not sure why you would inject it elsewhere. 

Probably warrants a investigation of the uboot-envtools package itself.

pprindeville commented on package/boot/uboot-envtools/files/layerscape:
```diff
@@ -0,0 +1,25 @@
+board=$(board_name)
+
+case "$board" in
+	traverse,ls1043v | \
```
You could also do: `traverse,ls1043[sv]` ...

pprindeville commented on package/boot/uboot-envtools/files/layerscape:
```diff
@@ -0,0 +1,25 @@
+
+. /lib/uboot-envtools.sh
+. /lib/functions.sh
+
```
No call to `board_config_update` necessary here?

pprindeville commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043s.dts:
```diff
@@ -0,0 +1,339 @@
+
+};
+
+/* The LS1043S does not expose the
```
Because?

mcbridematt replied:
It isn't possible to build a board that has both SATA and 10G ports (see https://traverse.com.au/docs/ls1043s/faq/). The device tree setting is to prevent the AHCI driver from enumerating.

pprindeville commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043v.dts:
```diff
@@ -0,0 +1,315 @@
+		reg = <0x4F>;
+		lltc,meas-mode = <6 3>;
+	};
+
```
Delete

pprindeville commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043v.dts:
```diff
@@ -0,0 +1,315 @@
+		reg = <0x24>;
+		gpio-base = <100>;
+	};
+
```
Delete

pprindeville commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043v.dts:
```diff
@@ -0,0 +1,315 @@
+	i2c-switch@70 {
+		compatible = "nxp,pca9540";
+		#address-cells = <1>;
+		#size-ceels = <0>;
```
Typo.

pprindeville commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043v.dts:
```diff
@@ -0,0 +1,315 @@
+
+};
+
+/* sfp_eth5: sfp-eth5 {
```
Why is this commented out?

pprindeville commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043v.dts:
```diff
@@ -0,0 +1,315 @@
+	/* NOR, NAND Flashes and FPGA on board */
+	ranges = <0x0 0x0 0x0 0x7e800000 0x00010000>;
+
+		nand@1,0 {
```
Over-indented.

pprindeville commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043v.dts:
```diff
@@ -0,0 +1,315 @@
+
+		phy-connection-type = "sgmii";
+		fixed-link {
+			speed = <1000>;
```
1000 for a 10G interface?

pprindeville commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043v.dts:
```diff
@@ -0,0 +1,315 @@
+			reg = <0x7>;
+		};
+	};
+
```
Delete

pprindeville commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043v.dts:
```diff
@@ -0,0 +1,315 @@
+};
+
+
+/* The LS1043V does not expose the
```
Because?

pprindeville commented on target/linux/layerscape/image/Makefile:
```diff
@@ -45,6 +47,16 @@ define Build/append-ls-dtb
+	./mkits-multiple-config.sh -o $@.its -A $(LINUX_KARCH) -v $(LINUX_VERSION) \
+		-k $@ -a $(KERNEL_LOADADDR) -e $(if $(KERNEL_ENTRY),$(KERNEL_ENTRY),$(KERNEL_LOADADDR)) \
+		-C gzip -c 1 -c 2 \
+		-d $(DEVICE_DTS_DIR)/freescale/traverse-ls1043s.dtb -D "Traverse_LS1043S" -n "ls1043s" -a $(FDT_LOADADDR) -c 1\
```
space before backslash

pprindeville commented on target/linux/layerscape/image/Makefile:
```diff
@@ -45,6 +47,16 @@ define Build/append-ls-dtb
+		-d $(DEVICE_DTS_DIR)/freescale/traverse-ls1043s.dtb -D "Traverse_LS1043S" -n "ls1043s" -a $(FDT_LOADADDR) -c 1\
+		-d $(DEVICE_DTS_DIR)/freescale/traverse-ls1043v.dtb -D "Traverse_LS1043V" -n "ls1043v" -a $(FDT_LOADADDR) -c 2
+	PATH=$(LINUX_DIR)/scripts/dtc:$(PATH) mkimage -f $@.its $@.new
+	@mv $@.new $@
```
I wouldn't bother hiding this line... but do you need a `-f`?

mcbridematt replied:
-f is for the FIT image source/ITS
./staging_dir/host/bin/mkimage [-D dtc_options] [-f fit-image.its|-f auto|-F] [-b <dtb> [-b <dtb>]] [-i <ramdisk.cpio.gz>] fit-image

pprindeville replied:
No, I mean mv-f

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+for (( index=1; index<=$MAX_IMG; index++ )); do
+	declare -a img$index
+	for i in {0..13}; do
+		eval img${index}[$i]=""
```
Why do you need `eval` here and not `let`?

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+for (( index=1; index<=$MAX_CONF; index++ )); do
+	declare -a conf$index
+	for i in {0..9}; do
+		eval conf${index}[$i]=""
```
Ditto

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+#	10: conf friendly name
+
+usage() {
+	echo "Usage: `basename $0` -A arch -v version -o its_file" \
```
Just use a "here document"...

```
cat <<__EOF__
...
__EOF__
```

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	local max_a=0
+	local max_i=0
+
+	if echo $1 | grep -q img; then
```
```
case "$1" in
*img*)
   ...
*conf*)
   ...
```

etc.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+		max_i=10
+		let a=$(echo $1 | awk -F "conf" '{print $2}')
+	fi
+	if [ ${a} -lt 0 -o ${a} -gt ${max_a} -o \
```
You don't need all these curly braces...

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	if [ -z "$4" ]; then
+		eval $1[$2]=$3
+	else
+		eval $1[$2]=\"\${$1[$2]} $3\"
```
You can't use `+=` here?

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+array_get()
+{
+	local val
+	eval val=\${$1[$2]}
```
Look at what you can do with `${!$1}` instead....

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+}
+
+parse_args() {
+	local i=-1 k=-1 d=-1 r=-1
```
I'd consider using:

```
local -i i=-1 k=-1 d=-1 r=-1
```

then you can do:

```
let i++
```

below.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+				array_put img$i 5 $OPTARG;;
+			c ) array_put img$i 7 $OPTARG append;;
+			D ) array_put img$i 8 $OPTARG;;
+			d ) i=$(($i + 1));
```
Don't need semis on any of these lines since you don't have multiple statements on the same line.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+				array_put img$i 10 $OPTARG;
+				;;
+			* ) echo "Invalid option passed to '$0' (options:$@)"
+			usage;;
```
Indent

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	case $1 in
+		signature) valid="sha-1,rsa-2048 sha-256,rsa-2048 sha-256,rsa-4096";;
+		compression) valid="gzip bzip2 none";;
+		hash) valid="sha1 md5 crc32";;
```
No catch-all?

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+
+#
+# sanity check for signature, compression and hash
+#
```
What are the arguments to this function?

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+		compression) valid="gzip bzip2 none";;
+		hash) valid="sha1 md5 crc32";;
+	esac
+	if ! echo $valid | grep -q "$2"; then
```
Not how I would do this, but maybe use `"\<$2\>"`

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	description = "U-Boot fitImage for ${VERSION} kernel";
+	#address-cells = <1>;
+EOF
+	;;
```
Indent `;;` to match rest of code.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+		echo -e "\n\timages {" >> ${OUTPUT};;
+	confstart)
+#		echo -e "\tconfigurations {\n\t\tdefault = \"conf@${2:-0}\";" \
+	echo -e "\tconfigurations {\n" \
```
Indent...  also why split this line?

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	local checksum=${6:-sha1}
+	local name=${7}
+
+	[ -z "${name}" ] || name=" ${name}"
```
```
name="${name:+ }$name"
```

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+
+	[ -z "${name}" ] || name=" ${name}"
+	cat << EOF >> ${OUTPUT}
+		kernel@${count} {
```
most of these curly braces are unnecessary.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+EOF
+	emit_cksum ${checksum}
+
+	if [ -z "$SIGN_IN_CONF" ] ; then
```
No space between `]` and `;`

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+		emit_signature "$9" "" "" "$8" "" ""
+	fi
+
+	echo "		};" >> ${OUTPUT}
```
```
echo -e "\t};" >> ${OUTPUT}
```

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+# Emit fdt node
+#
+emit_fdt() {
+	local image=${1}
```
Again most of these braces aren't necessary.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	local name=${5}
+	local loadaddr=${6}
+
+	[ -z "${name}" ] || name=" ${name}"
```
```
name="${name:+ }$name"
```

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+			compression = "none";
+EOF
+	emit_cksum ${checksum}
+	if [ -z "$SIGN_IN_CONF" ] ; then
```
As above.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	if [ -z "$SIGN_IN_CONF" ] ; then
+		emit_signature "" "$7" "" "" "$6" ""
+	fi
+	echo "		};" >> ${OUTPUT}
```
As above.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+# Emit ramdisk node
+#
+emit_ramdisk() {
+	local image=${1}
```
I'd drop most of these braces...

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	local checksum=${4:-sha1}
+	local name=${5}
+
+	[ -z "${name}" ] || name=" ${name}"
```
As above.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	local conf_csum="sha1"
+
+	config_name="conf@${1}"
+	if [ ! -z "${11}" ]; then
```
Always write `-n` and not `! -z`... plus you can use shorthand here:

```
[ -n "$11" ] && config_name="$11"
```

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	if [ ! -z "${11}" ]; then
+		config_name="${11}"
+	fi 
+	if [ -z "${2}" ]; then
```
```
[ -z "$2" ] && {
...
}
```

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+		fdt_line=""
+	else
+		conf_desc="Boot Linux kernel with FDT blob"
+		fdt_line="
```
I'd use `$'...'` here...

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+	# Test if we have any ROOTFS at all
+	if [ -n "${4}" ] ; then
+		conf_desc="$conf_desc + ramdisk"
+		fdt_line="${fdt_line}
```
Ditto.  Plus you don't need curly braces.

Curly braces are mostly useful for a complex interpolation like `${name:+...}` so you know where the end of the interpolation is, or else when concatenating strings to variable values like `${var}suffix` where you don't want to look up `$varsuffix` as a variable name.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+		emit_signature "$5" "$6" "$7" "$8" "$9" "${10}"
+	fi
+
+	echo "		};" >> ${OUTPUT}
```
Ditto.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+#
+remove_prefix_space()
+{
+	echo "$@" | sed "s:^ ::g"
```
`s:^ *::` ?

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+			img_conf=$(remove_prefix_space $img_conf)
+			img_hash=$(remove_prefix_space $img_hash)
+
+			[ "${img_type}" == $t ] || continue
```
`==` is a bashism

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+			esac
+
+			# set up configuration data
+			for img_c in ${img_conf}; do
```
Don't need curly braces...

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+				img_i=""
+				#set up default configuration if its not set
+				[ -n "$DEF_CONFIG" ] || DEF_CONFIG=$img_c
+				[ -n "${img_c}" ] || continue
```
You've already tested for non-empty here so the next for empty on the next line is redundant.

pprindeville commented on target/linux/layerscape/image/mkits-multiple-config.sh:
```diff
@@ -0,0 +1,594 @@
+				array_put conf$img_c $(($img_i + 6)) ${img_key}
+				array_put conf$img_c 10 $img_cname
+			done
+			img_index=$(($img_index + 1))
```
I'd use `local -e img_index` and `let img_index++` here...

pprindeville commented on target/linux/layerscape/patches-4.9/304-dts-layerscape-add-traverse-ls1043.patch:
```diff
@@ -0,0 +1,89 @@
+From c0612164b379ebc8964da6bc6f6ced9736dce488 Mon Sep 17 00:00:00 2001
```
Where is this patch coming from?  Was it upstreamed?  What's its status if it was?

mcbridematt replied:
This particular patch applies on top of NXPs patchset (in patches-4.9/), hence the number-sequence might be incorrect. Once layerscape is moved to 4.14, that should blow a lot of the uglyness away.

I haven't submitted our DTS to mainline yet, I think I should send the hwmon drivers up first so the bindings for those can be properly defined?

pprindeville commented on target/linux/layerscape/patches-4.9/820-rgmii-fixed-link.patch:
```diff
@@ -0,0 +1,62 @@
+ 
++    if (ENET_INTERFACE_FROM_MODE(p_Memac->enetMode) == e_ENET_IF_RGMII) {
++	if (mac_dev->fixed_link) {
++		printk(KERN_INFO "This is a fixed-link, forcing speed %d duplex %d\n",mac_dev->fixed_link->speed,mac_dev->fixed_link->duplex);
```
I'd have used comma-space but since it's already that way upstream...

pprindeville reviewed (changes requested):
New platforms are particularly hard to get accepted so I went over this with a fine-toothed comb...

mcbridematt commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043s.dts:
```diff
@@ -0,0 +1,339 @@
+		status = "okay";
+		phy-connection-type = "sgmii";
+		fixed-link {
+			speed = <1000>;
```
The SFP+ slot can be set to operate at 1G (with a different power-on config).

If the speed is set to 10000 in DTS the DPAA driver doesn't seem to work at 1G, if it is set to 1000 both cases work.

mcbridematt commented on target/linux/layerscape/base-files/lib/preinit/05_layerscape_reorder_eth:
```diff
@@ -0,0 +1,30 @@
+	"traverse,ls1043v" | \
+	"traverse,ls1043s")
+
+	# Reorder ethernet interfaces to match the physical order
```
Our u-boot will set the MAC addresses in the 'correct' order so something could use that to reindex them.
Another thought is that we could add something into the DTS to set the 'preferred' name. I think there is something like this for ACPI already.

mcbridematt commented on target/linux/layerscape/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,21 @@
+		ucidef_set_interface_wan "eth4"
+		;;
+	traverse,ls1043s)
+		ucidef_set_interface_lan "eth0 eth1 eth2 eth3 eth6"
```
eth5 is for the VDSL (on board on the 'V, future add in card on the 'S).

mcbridematt commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043s.dts:
```diff
@@ -0,0 +1,339 @@
+		};
+	}; 
+};
+/* Only for Rev 1 boards and later
```
The UHS/DDR modes don't work correctly on the current hardware revision, once Rev 1 comes along and I get all the Rev '0' boards back, I'll send a change to enable them.

mcbridematt commented on target/linux/layerscape/files/arch/arm64/boot/dts/freescale/traverse-ls1043s.dts:
```diff
@@ -0,0 +1,339 @@
+		#gpio-cells = <2>;
+		reg = <0x24>;
+		gpio-base = <100>;
+		/* 
```
I was hoping we could use this to 'name' GPIOs rather than /etc/config/system, unfortunately GPIO-hog means they can't be re-exported by userspace. I'll get rid of them.

mcbridematt commented on package/boot/uboot-envtools/files/layerscape:
```diff
@@ -0,0 +1,25 @@
+
+. /lib/uboot-envtools.sh
+. /lib/functions.sh
+
```
I think that gets done by the parent process (which goes over /etc/uci-defaults/)?

(This is not /etc/board.d)

mcbridematt commented:
Thanks for the feedback - I should have paid a bit more attention to the DTS and upgrade script.

The upgrade script has been cleaned up a bit, I moved our method into platform_do_upgrade_traverse_nandubi() so it doesn't interfere with any other future implementations.

pprindeville commented:
Let me know when you have pushed a new version.

mcbridematt committed:
layerscape: add Traverse LS1043-S support

The Traverse LS1043-S board is a router board based on
NXP/Freescale's LS1043 SoC, with 4x1GBase-T, 1 SFP and 1 SFP+,
as well as miniPCIe and M.2 LTE.

Unlike the Layerscape reference boards, the LS1043-S board has
NAND flash and uses the mainline U-Boot.

This patch implements support for the LS1043-S board, as well as
the earlier LS1043-V board. It is our intention that all boards
in this family (LS1043-S and later, Five64) will boot the same binary.

Not included in this patchset are the hwmon drivers not in the kernel
(emc1704,pac1934) or the bootloader.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
layerscape: add multiple-configuration generator for FIT images

This is required for the Traverse LS1043 family, we generate a FIT image
that works on all boards across the family. This is done by creating
multiple configurations that select the right DTB for the board.

The bootloader on these boards is configured to boot like this:
bootm $kernel_load#ls1043s
bootm $kernel_load#ls1043v

This is based on earlier work by Jason Wu for Zynq:
https://lists.openwrt.org/pipermail/openwrt-devel/2016-March/040460.html

Modified to add FDT load addresses and multiple configurations with DTB.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt commented on target/linux/layerscape/base-files/etc/board.d/03_gpio_switches:
```diff
@@ -0,0 +1,28 @@
+	ucidef_add_gpio_switch "lte_reset" "LTE Reset" "377"
+	ucidef_add_gpio_switch "lte_disable" "LTE Airplane mode" "378"
+	;;
+traverse,ls1043s)
```
gpio-hog is the closest you can get in the mainline kernel, but the 'hog' part is an issue.
SFP control I hope to handle with the recently added sfp-bus/phylink, but it looks like this requires some support in the network drivers.

mcbridematt commented:
I've pushed a new one, as mentioned the DTS and upgrade is sorted.

uboot-envtools is a bit funny - every other platform handler has the same structure - I'm not saying that makes it right, though.

mcbridematt commented:
I appear to have missed some of the comments on the ITS generator - I'll go have another look at it.
It is overkill for the use case (we're just trying to add multiple DTBs and start configurations) but my script-fu isn't at the level of trying to adopt the existing mkits.sh or rolling my own.

pprindeville reviewed (commented):
No further issues.  I'll let someone with better DTS and ARM-fu do the approval, though.

blogic commented:
merged, Thanks !

yangbolu1991 commented:
Hi John and Mathew,

Sorry for missing this pull request.
Do you think whether it will be better to create a new target of Traverse to maintian such board?
The NXP Layerscape target was for some reference design boards of NXP.
The Traverse LS1043-S board in this pull request was a router board from Traverse.

For NXP Layerscape target, we usually update kernel patches, u-boot and releated packages half a year according to latest NXP LSDK release. We have system test for each board before sending out patches to OpenWrt community.
I'm not sure whether this Traverse LS1043-S board would always use same kernel/u-boot/firmware/packages with NXP Layerscape target. 
So I need some suggestions.

Thanks a lot.

mcbridematt commented:
Understood.
Our board target is configured to not build the NXP u-boot,rcw,ppa etc., we expect our customers will update this separately according to our directions.

Kernel updates should be ok, I don't think there are any changes in the latest LSDK that will cause an issue for us.

yangbolu1991 commented:
Thanks a lot for clarifying this, Mathew.
BTW, did you use 32-bit mode for this board? I just got a compile issue within my pull request.
I'm not sure whether my patches introduced this issue. Or it was there when you sent your pull request.
Anyway I will have a look tomorrow :)

cp /home/lyb/projects/openwrt/DN-upstream/openwrt/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-layerscape_armv8_32b/Image /home/lyb/projects/openwrt/DN-upstream/openwrt/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-layerscape_armv8_32b/traverse-five64-kernel.itb

mcbridematt commented:
I only intended to enable 64-bit mode, it looks like our target definition (https://github.com/openwrt/openwrt/blob/master/target/linux/layerscape/image/Makefile#L198) should have be placed inside the 'ifeq ($(SUBTARGET),armv8_64b)' block with ls1088/ls2088ardb. 

I will send a pull request to resolve, thanks for letting me know.

yangbolu1991 commented:
Thanks for confirmation, Mathew.
Let me do this. I'm going to split the image/Makefile per subtarget. We will add armv7 too.
